source/bluetooth/MicroBitUtilityService.cpp

```diff
@@ -0,0 +1,385 @@
+     */
+    void onReplySent( const void *data)
+    {
+        reply_t *reply = (reply_t *) data;
+        if ( ( reply->job & 0x0F) != jobLowERR)
+        {
+            jobLow++;
+            if ( jobLow > jobLowMAX)
+                jobLow = 0;
+        }
+    }
+};
+
+
+/**
+ * Return a pointer to the shared service, creating it if necessary
+ * @param _ble an instance of BLEDevice
+ * @param _messageBus An instance of a MessageBus to interface with.
+ * @param _storage A persistent storage manager to use to hold non-volatile state.
+ * @return a pointer to the service, or NULL if it has not been created
+ */
+MicroBitUtilityService *MicroBitUtilityService::createShared( BLEDevice &_ble, EventModel &_messageBus, MicroBitStorage &_storage, MicroBitLog &_log)
+{
+    if ( !shared)
+        shared = new MicroBitUtilityService( _ble, _messageBus, _storage, _log);
+    return shared;
+}
+
+/**
+ * Constructor.
+ * Create a representation of the Utility Service
+ * @param _ble The instance of a BLE device that we're running on.
+ * @param _messageBus An instance of a MessageBus to interface with.
+ * @param _storage A persistent storage manager to use to hold non-volatile state.
+ * @param _log An instance of a MicroBitLog to interface with.
+ */
+MicroBitUtilityService::MicroBitUtilityService( BLEDevice &_ble, EventModel &_messageBus, MicroBitStorage &_storage, MicroBitLog &_log) :
+    messageBus(_messageBus), storage(_storage), log(_log), workspace(NULL)
+{
+    // Initialise data
+    memclr( characteristicValue, sizeof( characteristicValue));
+
+    // Register the base UUID and create the service.
+    RegisterBaseUUID( bs_base_uuid);
+    CreateService( serviceUUID);
+
+    CreateCharacteristic( mbbs_cIdxCTRL, charUUID[ mbbs_cIdxCTRL],
+                         characteristicValue,
+                         0, sizeof(characteristicValue),
+                         microbit_propWRITE | microbit_propWRITE_WITHOUT | microbit_propNOTIFY);
+    
+    if ( getConnected())
+        listen(true);
+}
+
+
+/**
+ * Invoked when BLE connects.
+ * Start listening for events
+ */
+void MicroBitUtilityService::onConnect( const microbit_ble_evt_t *p_ble_evt)
+{
+    listen(true);
+}
+
+
+/**
+ * Invoked when BLE disconnects.
+ * Stop listening for events
+ */
+void MicroBitUtilityService::onDisconnect( const microbit_ble_evt_t *p_ble_evt)
+{
+    listen(false);
+}
+
+
+/**
+ * Callback. Invoked when any of our attributes are written via BLE.
+ * Record the  request, and trigger processing.
+ * Requests are not queued. If a request is actively being processed, ignore the new request, otherwise simply override the previous one.
+ */
+void MicroBitUtilityService::onDataWritten(const microbit_ble_evt_write_t *params)
+{
+    if ( params->handle == valueHandle( mbbs_cIdxCTRL) && params->len > 0 && params->len < sizeof(request_t))
+    {
+        if ( workspace)
+        {
+            if ( workspace->lock)
```

Comment:
is this safe @martinwork? If I understand correctly, this code will silently ignore incoming requests if a request is already being processed? Is that conventional for BLE apps? Seems quite lossy?

Reply:
I believe my thinking was that the each request would expect a reply (both current ones do), and the client wouldn't send multiple requests.

Reply:
I think that's ok - we just need to record somewhere in docs that we don't permit overlapping transactions.